freyja/read_analysis_tools.py

```diff
@@ -388,6 +400,360 @@ def filter(query_mutations, input_bam, min_site, max_site, output, refname):
+                    ins_offset - del_offset - 1
+                read_end = (locus - start) - codon_position + \
+                    ins_offset - del_offset + 2
+
+                alt_codon = MutableSeq(seq[read_start:read_end])
+
+                if len(alt_codon) % 3 != 0 or len(alt_codon) == 0:
+                    continue  # Possible fail case: codon spans multiple reads
+                alt_aa = alt_codon.translate()
+                if alt_aa == 'X':
+                    continue
+                aa_mut = f'{snp}({gene}:{ref_aa}{aa_locus}{alt_aa})'
+
+                muts_final.append(aa_mut)
+            if coverage_start is None or start < coverage_start:
+                coverage_start = start
+            if coverage_end is None or start + len(seq) > coverage_end:
+                coverage_end = start + len(seq)
+
+        if spans_region:
+            if coverage_start > min_site or coverage_end < max_site:
+                continue
+
+        muts_final = sorted(list(set(muts_final)), key=nt_position)
+        name = ' '.join([str(mut).replace(' ', '') for mut in muts_final])
+        if len(name) > 1:
+            if name not in co_muts:
+                co_muts[name] = 1
+            else:
+                co_muts[name] += 1
+            coverage[name] = (coverage_start, coverage_end)
+    samfile.close()
+
+    df = pd.DataFrame()
+    df['Covariants'] = [k for k in co_muts]
+    df['Count'] = [co_muts[k] for k in co_muts]
+    df['Frequency'] = df['Count'] / all_reads_count
+    df['Coverage_start'] = [coverage[k][0] for k in co_muts]
+    df['Coverage_end'] = [coverage[k][1] for k in co_muts]
+
+    df = df[df['Count'] >= min_count]
+
+    # Sort by site of first mutation
+    df['sort_col'] = [nt_position(s.split(' ')[0]) for s in df.Covariants]
+    df = df.sort_values('sort_col').drop(labels='sort_col', axis=1)
+
+    df.to_csv(output, sep='\t', index=False)
+    print(f'covariants: Output saved to {output}')
+    return df
+
+
+def plot_covariants(covar_file, output, min_mutations):
+
+    # Define columns (mutations in samples)
+    covars = pd.read_csv(covar_file, sep='\t', header=0)
+
+    # Get covariants and unique mutations
+    nt_muts = []
+    patterns = []
+    for c in covars.iloc[:, 0]:
+        sample = c.split(' ')
+        for mut in sample:
+            if mut not in nt_muts:
+                nt_muts.append(mut)
+        patterns.append(sample)
+
+    coverage_start = [int(i) for i in covars.iloc[:, 3]]
+    coverage_end = [int(i) for i in covars.iloc[:, 4]]
+    coverage_ranges = {str(patterns[i]): (
+        coverage_start[i], coverage_end[i]) for i in range(len(patterns))}
+
+    nt_muts = sorted(nt_muts, key=nt_position)
+    colnames = []
+    sites = {}
+    for mut in nt_muts:  # remove duplicates
+        nt_site = nt_position(mut)
+        if ',' in mut:
+            aa_site = mut.split(')(')[1][:-1]
+        else:
+            aa_site = mut.split('(')[1][:-1]
```

Comment:
Dropping the nt mutation info will make it hard to spot nonsynonymous mutations. Can we add a switch to the cli that will turn of this splitting step?